File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第8章数的拆分.java
//数的拆分    lanqiaoOJ 2090 

import java.io.*;
import java.util.*;

public class Main {

    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static Scanner cin = new Scanner(System.in);
    static int cnt = 0 , N = 4000;
    static int[]prime = new int[4001];
    static int[]vis = new int[4001];
    public static void init(){
        for(int i = 2; i <= N; i++){
            if(0 == vis[i]) prime[++cnt] = i;
            for(int j = i*i; j <= N; j += i)   vis[j] = 1;
        }
    }
    public static boolean square_number(long x)
    {
        long y = (long) Math.sqrt(x);
        while(y * y <= x)//开平方存在精度误差，需要多判断几步
        {
            if(y * y == x)return true;
            y += 1;
        }
        return false;
    }
    public static boolean cubic_number(long x)
    {
        long y = (long) Math.pow(x, 1.0 / 3);
        while(y * y * y <= x)//开立方存在精度误差，需要多判断几步
        {
            if(y * y * y == x)return true;
            y += 1;
        }
        return false;
    }
    public static void main(String[] args) throws IOException {
        init();
        int T = cin.nextInt();
        while(T -- > 0){
            long a = cin.nextLong();
            if(square_number(a) || cubic_number(a)) {
                System.out.println("yes");
                continue;
            }
            boolean flag = true;
            for(int i = 1 ; i <= cnt ; i ++) if(a % prime[i] == 0){
                int mi = 0;
                while(a % prime[i] == 0) {a /= prime[i] ; mi ++;}
                if(mi == 1) {
                    flag = false;
                    break;
                }
            }
            if(flag && (square_number(a) || cubic_number(a))) System.out.println("yes");
            else System.out.println("no");
        }
    }

}

File: 配套资源/配套代码/程序设计竞赛专题挑战教程-代码/第8章数的拆分.py
#数的拆分   lanqiaoOJ 2090
 

from math import *
N = 4000

prime = [0]*N
vis = [0]*N
cnt = 0
def E_sieve():
    global cnt
    for i in range(2,N):
        if not vis[i]: cnt+=1; prime[cnt] = i
        for j in range(i*i,N,i):    vis[j] = 1

def solve():
    a = int(input())
    for i  in range(1,cnt+1):
        c = 0
        while a % prime[i] == 0: a/=prime[i]; c+=1
        if c==1: print("no"); return
    
    k = int(sqrt(a))
    if k*k == a: print("yes"); return  #检查n是否为平方数
    k = int(pow(a, 1/3))
    if k*k*k==a:  print("yes"); return #检查n是否为立方数
    print("no")

E_sieve()
T=int(input())
for i in range(T): solve()
